lib/service/storage/virtualIndex.ts

```diff
@@ -0,0 +1,190 @@
+import { Service } from "../service";
+
+export class VirtualIndex extends Service {
+  public static createEvent = "core:storage:virtual-index:cache:create";
+  public static deleteEvent = "core:storage:virtual-index:cache:delete";
+
+  public static createVirtualIndexEvent =
+    "core:storage:virtual-index:create:after";
+
+  public static deleteVirtualIndexEvent =
+    "core:storage:virtual-index:delete:after";
+
+  constructor() {
+    super("VirtualIndex", global.kuzzle.config.services.storageEngine);
+  }
+
+  public virtualIndexMap: Map<string, string> = new Map<string, string>(); //Key : virtual index, value : real index
+
+  async init(): Promise<void> {
+    await this.buildCollection();
+    await this.initVirtualIndexList();
+
+    global.kuzzle.onAsk(VirtualIndex.createEvent, (physical, virtual) =>
+      this.addInVirtualIndexMap(physical, virtual)
+    );
+
+    global.kuzzle.onAsk(VirtualIndex.deleteEvent, (info) =>
+      this.removeInVirtualIndexMap(info)
+    );
+  }
+
+  addInVirtualIndexMap(physical, virtual) {
+    this.virtualIndexMap.set(virtual, physical);
+  }
+
+  removeInVirtualIndexMap(notification) {
+    this.virtualIndexMap.delete(notification);
+  }
+
+  getPhysicalIndex(name: string): string {
+    if (this.virtualIndexMap.has(name)) {
+      return this.virtualIndexMap.get(name);
+    }
+    return name;
+  }
+
+  isVirtual(name: string): boolean {
+    return this.virtualIndexMap.has(name);
+  }
+
+  getId(index: string, id: string): string {
+    if (this.isVirtual(index)) {
+      return index + id;
+    }
+    return id;
+  }
+
+  randomString(size = 20) {
+    const buf = Buffer.alloc(size);
+    for (let i = 0; i < size; i++) {
+      buf.writeUInt8((Math.random() * 255) & 255, i);
+    }
+    return buf.toString("base64").slice(0, size);
+  }
+
+  getVirtualId(index: string, id: string): string {
+    if (this.isVirtual(index) && id.startsWith(index)) {
+      return id.substring(index.length, id.length);
+    }
+    return id;
+  }
+
+  public async createVirtualIndex(virtualIndex: string, index: string) {
+    global.kuzzle.emit(VirtualIndex.createVirtualIndexEvent, {
```

Comment:
I think the event should be dispatched after the successful creation in the `virtual-indexes` collection top avoid unconsistent states